creme/optim/averager.py

```diff
@@ -0,0 +1,83 @@
+import collections
+
+from . import base
+from . import schedulers
+
+
+__all__ = ['Averager']
+
+
+class Averager(base.Optimizer):
+    """An Stochastic weight averaging wrapper around various sgd optimzers.
+
+    Example:
+
+        ::
+
+            >>> from creme import linear_model
+            >>> from creme import metrics
+            >>> from creme import model_selection
+            >>> from creme.optim.schedulers import Cyclic
+            >>> from creme.optim import Adam, Averager
+            >>> from creme import preprocessing
+            >>> from creme import stream
+            >>> from sklearn import datasets
+
+            >>> X_y = stream.iter_sklearn_dataset(
+            ...     dataset=datasets.load_breast_cancer(),
+            ...     shuffle=True,
+            ...     random_state=42
+            ... )
+
+            >>> adam = Adam(lr=Cyclic(1, 0.01, 100))
+            >>> optimizer = Averager(adam)
+            >>> model = (
+            ...     preprocessing.StandardScaler() |
+            ...     linear_model.LogisticRegression(optimizer)
+            ... )
+            >>> metric = metrics.F1()
+
+            >>> model_selection.progressive_val_score(X_y, model, metric)
+            F1: 0.959441
+
+    References:
+        1. `Averaging Weights Leads to Wider Optima and Better Generalization
+            <https://arxiv.org/abs/1803.05407>`_
+
+    """
+
+    def __init__(self, base_optim):
+        self.base_optim = base_optim
+        self.n_iterations = 0
+        self.c = (
```

Comment:
What is `c`?

Reply:
`c` is the cycle length of the `Cyclic lr` scheduler. If the `base_optim` is using an `Cyclic lr` then `Averager` will average the weights in interval of length `c` and if `base_optim` is not using `Cyclic lr` than `c=1` which means `Averager` will average the weights after every iteration.

Actually in the mentioned paper and blogs I read on Averaging, they used a `Cyclic scheduler` for `lr`. That's why I add this.